### PR TITLE
Implement API priorities (writing)

### PR DESCRIPTION
**Pull request by NobodysNightmare**
## Openproject work package

https://community.openproject.org/work_packages/17409
## Notes
- This PR ~~completely~~ partially contains https://github.com/opf/openproject/pull/2456

### CODE REVIEW

**Comment by TeatroIO** [comment]
I've prepared a stage to preview changes. <a href='http://feature-api-write-priorities.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://feature-api-write-priorities.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

File: lib/api/v3/work_packages/form/work_package_schema_representer.rb [hunk 1]
```diff
@@ -91,6 +91,19 @@ class WorkPackageSchemaRepresenter < ::API::Decorators::Single
...
                    }
 
+          property :priority,
+                   exec_context: :decorator,
+                   getter: -> (*) {
+                     priority_origin = represented
+
+                     if represented.persisted? && represented.priority_id_changed?
+                       priority_origin = represented.class.find(represented.id)
+                     end
```
Comment (ulferts) [inline lib/api/v3/work_packages/form/work_package_schema_representer.rb]: @NobodysNightmare, @Azure7111 

This logic, getting the priorities that are assignable before the change in priority, is probably not needed. There is no change in assignable_priorities when the priority is changed.

Having read through the code I wrote above, the same applies to the version although that will have to be revised once the project can be changed. But that is true for other fields as well.

**Comment by ulferts** [comment]
@Azure7111, @NobodysNightmare

When updating with a botched link to a priority e.g. /api/v3/prioritiesss/4 an untranslated error is returned:
`For property 'translation missing: en.attributes.priority' a resource of type 'translation missing: en.attributes.priority' is expected but...`

**Comment by ulferts** [comment]
Updating the form and the WP itself works. However, I had to test that manually as there are no integrative tests. :point_up_2:

**Comment by ulferts** [comment]
There is another issue revealed here that is not limited to priorities but seems to be a general issue. If a non existing ID is provided. e.g. 0. The error returned is that the field can't be blank. That seems to be wrong.

**Comment by NobodysNightmare** [comment]
> There is another issue revealed here that is not limited to priorities but seems to be a general issue. If a non existing ID is provided. e.g. 0. The error returned is that the field can't be blank. That seems to be wrong.

Yeah, that is a known limitation. To be honest: I pointed that out once, but then I was told to believe that this is an error that is out of scope because it also affects non-API parts. I therefore accepted it and agreed that _someone_ should fix it later.

I hope I did not become _someone_...

File: lib/api/v3/work_packages/form/work_package_schema_representer.rb [hunk 2]
```diff
@@ -81,16 +81,17 @@ class WorkPackageSchemaRepresenter < ::API::Decorators::Single
           property :version,
                    exec_context: :decorator,
                    getter: -> (*) {
-                     version_origin = represented
-
-                     if represented.persisted? && represented.fixed_version_id_changed?
-                       version_origin = represented.class.find(represented.id)
-                     end
-
-                     SchemaAllowedVersionsRepresenter.new(version_origin.assignable_versions,
```
Comment (ulferts) [inline lib/api/v3/work_packages/form/work_package_schema_representer.rb:90]: Thanks

**Pull request merged**